File: app/src/main/java/com/example/exitpro/OTPVerification.java
package com.example.exitpro;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import android.app.ProgressDialog;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONException;
import org.json.JSONObject;

public class OTPVerification extends Fragment {

    private EditText otpEditText;
    private Button verifyButton;
    private ProgressDialog progressDialog;
    private RequestQueue requestQueue;
    String guardId;
    String otpURL = "https://exitpro-backend.onrender.com/security/otpMatch";

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_o_t_p_verification, container, false);

        otpEditText = rootView.findViewById(R.id.editText_otp);
        verifyButton = rootView.findViewById(R.id.button_verify_otp);
        requestQueue = Volley.newRequestQueue(getActivity());

        Bundle bundle = getArguments();
        if (bundle != null) {
            guardId = bundle.getString("Guard ID");
        }
        verifyButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Call method to verify OTP
                verifyOTP();
            }
        });

        return rootView;
    }

    private void verifyOTP() {
        // Implement OTP verification logic here
        // For example, you can compare the entered OTP with the OTP received from the backend
        // If the OTP matches, you can proceed with the login process
        // Otherwise, display an error message to the user
        showLoadingDialog();
        String otp = otpEditText.getText().toString();
        JSONObject jsonBody = new JSONObject();
        try {
            jsonBody.put("guardId", guardId);
            jsonBody.put("otp", otp);
            Log.e("SAYAK",jsonBody.toString());
        } catch (JSONException e) {
            e.printStackTrace();
        }

        JsonObjectRequest request = new JsonObjectRequest(
                Request.Method.POST,
                otpURL,
                jsonBody,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        dismissLoadingDialog();
                        try {
                            boolean isSuccess = response.getBoolean("isSuccess");
                            if (isSuccess) {
                                saveAccessToken(otp);
                                Intent intent = new Intent(getActivity(), HomeActivity.class);
                                startActivity(intent);
                                getActivity().finish();
                            } else {
                                dismissLoadingDialog();
                                otpEditText.setError("Wrong OTP");
                                Toast.makeText(getActivity().getApplicationContext(), "Wrong OTP!",Toast.LENGTH_SHORT);
                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                            otpEditText.setError("Wrong OTP");
                            Toast.makeText(getActivity().getApplicationContext(), "Wrong OTP!",Toast.LENGTH_SHORT);
                            Log.e("SAYAK","catch block otpmatch");
                        }
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        dismissLoadingDialog();
                        Toast.makeText(getActivity(), "Error: " + error.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
        requestQueue.add(request);

    }
    private void saveAccessToken(String accessToken) {
        // Save access token securely (e.g., using SharedPreferences)
        SharedPreferences sharedPreferences = getActivity().getSharedPreferences("MyPrefs", Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString("access_token", accessToken);
        editor.apply();
    }
    private void showLoadingDialog() {
        progressDialog = new ProgressDialog(getActivity());
        progressDialog.setCancelable(false);
        progressDialog.setMessage("Checking OTP...");
        progressDialog.show();
    }

    private void dismissLoadingDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }

}

File: app/src/main/java/com/example/exitpro/fragment/OTPVerification.kt
package com.example.exitpro.fragment

import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.example.exitpro.R
import com.example.exitpro.activity.HomeActivity
import com.example.exitpro.data.api.ApiResponse
import com.example.exitpro.data.api.RetrofitClient
import com.example.exitpro.data.repository.ExitProRepository
import kotlinx.coroutines.launch

class OTPVerification : Fragment() {
    // UI elements
    private lateinit var otpEditText: EditText
    private lateinit var verifyButton: Button
    private var loadingDialog: Dialog? = null
    private val repository by lazy { ExitProRepository(RetrofitClient.apiService) }
    private var guardId: String? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_o_t_p_verification, container, false)

        // Initialize UI elements
        otpEditText = rootView.findViewById(R.id.editText_otp)
        verifyButton = rootView.findViewById(R.id.button_verify_otp)

        // Retrieve guard ID from arguments
        guardId = arguments?.getString("Guard ID")

        // Set click listener for the verify button using a lambda expression
        verifyButton.setOnClickListener { verifyOTP() }

        return rootView
    }

    /**
     * Verifies the OTP entered by the user.
     */
    private fun verifyOTP() {
        showLoadingDialog()

        val otp = otpEditText.text.toString()
        val currentGuardId = guardId

        if (currentGuardId == null) {
            dismissLoadingDialog()
            showError("Guard ID not found")
            return
        }

        if (otp.isEmpty()) {
            dismissLoadingDialog()
            showError("Please enter OTP")
            return
        }

        lifecycleScope.launch {
            when (val response = repository.verifyOTP(currentGuardId, otp)) {
                is ApiResponse.Success -> {
                    dismissLoadingDialog()
                    if (response.data.isSuccess) {
                        val guardName = response.data.guardName ?: "Guard"
                        saveAccessToken(otp, guardName)
                        val intent = Intent(activity, HomeActivity::class.java)
                        startActivity(intent)
                        activity?.finish()
                    } else {
                        showError("Wrong OTP")
                    }
                }
                is ApiResponse.Error -> {
                    dismissLoadingDialog()
                    showError("Error: ${response.message}")
                }
                is ApiResponse.Exception -> {
                    dismissLoadingDialog()
                    val errorMessage = when (response.exception) {
                        is java.net.UnknownHostException -> "No internet connection"
                        is java.net.SocketTimeoutException -> "Connection timeout"
                        else -> "Error: ${response.exception.message}"
                    }
                    showError(errorMessage)
                }
                is ApiResponse.Loading -> {
                    // Loading state already handled
                }
            }
        }
    }

    /**
     * Saves the access token securely using SharedPreferences.
     *
     * @param otp The access token to save.
     */
    private fun saveAccessToken(otp: String, guardName: String) {
        val sharedPreferences = activity?.getSharedPreferences("MyPrefs", Context.MODE_PRIVATE)
        sharedPreferences?.edit()?.apply {
            putString("otp", otp)
            putString("guard_name", guardName)
            apply()
        }
    }

    /**
     * Display a loading dialog while verifying OTP.
     */
    private fun showLoadingDialog() {
        loadingDialog = Dialog(requireContext()).apply {
            requestWindowFeature(Window.FEATURE_NO_TITLE)
            setContentView(R.layout.loading_dialog)
            setCancelable(false)
            window?.setBackgroundDrawableResource(android.R.color.transparent)
            show()
        }
    }

    /**
     * Dismiss the loading dialog if currently showing.
     */
    private fun dismissLoadingDialog() {
        loadingDialog?.takeIf { it.isShowing }?.dismiss()
    }

    /**
     * Shows an error message in the OTP edit text and as a toast.
     *
     * @param message The error message to show.
     */
    private fun showError(message: String) {
        otpEditText.error = message
        Toast.makeText(requireActivity().applicationContext, message, Toast.LENGTH_SHORT).show()
    }

}
